draft/ZEP0002.md

```diff
@@ -0,0 +1,285 @@
+---
+layout: default
+title: ZEP0002
+description: This ZEP proposes to add a sharding storage transformer to
+  the Zarr specification version 3. This allows 
+parent: draft ZEPs
+nav_order: 1
+---
+
+# ZEP 2 - Sharding storage transformer
+
+Authors:
+* Jonathan Striebel ([@jstriebel](https://github.com/jstriebel)), scalable minds
+
+Status: Draft
+
+Type: Specification
+
+Created: 2022-07-29
+
+This ZEP is accompanied by the sharding extension specification [[SHARDINGSPEC](#ref-SHARDINGSPEC)].
+A public review of the specification is done in PR **TODO**.
+
+
+## Abstract
+
+This ZEP proposes a new major version of the core specification which
+defines how to store and retrieve Zarr data. The ZEP also proposes a
+new framework for modularising specifications of codecs, storage
+systems and extensions.
+
+
+## Motivation and Scope
+
+Currently, Zarr maps one array chunk to one storage key. This becomes a bottleneck as storing large
+In many cases it becomes inefficient or impractical to store a large number of chunks directly as
+files or objects due to the design constraints of the underlying storage.
+For example the file block size and maximum inode number restrict the usage of numerouos small files
+for typical file systems, also cloud storage such as S3, GCS, and various distributed filesystems
+do not efficiently handle large numbers of small files or objects.
+
+Increasing the chunk size works only up to a certain point, as chunk sizes need to be small for
+read and write efficiency requirements, for example to stream data in browser-based visualization software.
+
+Therefore, chunks may need to be smaller than the minimum size of one storage key.
+In those cases it is efficient to store objects at a more coarse granularity than reading chunks.
+
+**Sharding solves this by allowing to store multiple chunks in one storage key, which is called a shard**:
+
+![sharding schema](../assets/images/sharding.png)
+
+
+The following example illustrates the need for sharding:
+
+The dataset has a shape of `(25000, 18000, 6000)` and a dtype of `uint8`. This makes for a total size of 2.4TB. For efficient reads and fast streaming of the data, e.g. to a browser-based interface, chunk sizes should be of max. `(64, 64, 64)`. For this dataset that would result in 10.3M chunks, and withouth sharding also to 10.3M files. Considering even many of these datasets, the number of files becomes impractical. With sharding, one could use shards of size `(2048, 2048, 2048)` which would each contain `(32, 32, 32)` chunks. That would reduce the number of files to ~300, while maintaining a practical write granularity.
```

Comment:
```suggestion
The dataset has a shape of `(25000, 18000, 6000)` and a dtype of `uint8`. This makes for a total size of 2.4TB. For efficient reads and fast streaming of the data, e.g. to a browser-based interface, chunk sizes should be of max. `(64, 64, 64)`. For this dataset that would result in 10.3M chunks, and without sharding also to 10.3M files. Considering even many of these datasets, the number of files becomes impractical. With sharding, one could use shards of size `(2048, 2048, 2048)` which would each contain `(32, 32, 32)` chunks. That would reduce the number of files to ~300, while maintaining a practical write granularity.
```

Reply:
Wird die "shard size" in voxel oder in chunks angeben?

Reply:
In der config gibt es einfach nur `chunks_per_shard`, in dem Beispiel hab ich es jetzt klarer formuliert.